File: ts/sb/struct_sim_info.ts
import * as _ from "./_"
import * as TplEnum from "./enum"

export interface SimInfo extends _.Serializable, _.Deserializable {
    id: number;
    title: string;
    content: string;
    a: boolean;
    b: boolean;
    c: boolean;
    d: boolean;
    zip: Uint8Array;
}

export const newSimInfo = (): SimInfo => {
    const s = {
        id: 0,
        title: "",
        content: "",
        a: false,
        b: false,
        c: false,
        d: false,
        zip: new Uint8Array(0),
    } as any as SimInfo;
    s.set = (buf: _.Buffer) => setSimInfo(buf, s);
    s.get = (buf: _.Buffer) => {
        const [res, err] = getSimInfo(buf);
        if (err === null) Object.assign(s, res);
        return err;
    };
    return s;
}

export const eqSimInfo = (a: SimInfo, b: SimInfo): boolean => {
    if (a === b) return true;
    if (a === null || b === null) return false;
    if (!_.eqU32(a.id, b.id)) return false;
    if (!_.eqText(a.title, b.title)) return false;
    if (!_.eqText(a.content, b.content)) return false;
    if (!_.eqBool(a.a, b.a)) return false;
    if (!_.eqBool(a.b, b.b)) return false;
    if (!_.eqBool(a.c, b.c)) return false;
    if (!_.eqBool(a.d, b.d)) return false;
    if (!_.eqBin(a.zip, b.zip)) return false;
    return true;
}

export const getSimInfo = (buf: _.Buffer): [SimInfo, Error | null] => {
    const s = newSimInfo();
    const bitmaskSize = 1;
    const [bits, err] = buf.read(bitmaskSize);
    if (err !== null) return [s, err];
    if (_.GetBit(bits, 0)) {
        const [v, err] = _.getU32(buf);
        if (err !== null) return [s, err];
        s.id = v;
    }
    if (_.GetBit(bits, 1)) {
        const [v, err] = _.getText(buf);
        if (err !== null) return [s, err];
        s.title = v;
    }
    if (_.GetBit(bits, 2)) {
        const [v, err] = _.getText(buf);
        if (err !== null) return [s, err];
        s.content = v;
    }
    s.a = _.GetBit(bits, 3);
    s.b = _.GetBit(bits, 4);
    s.c = _.GetBit(bits, 5);
    s.d = _.GetBit(bits, 6);
    if (_.GetBit(bits, 7)) {
        const [v, err] = _.getBin(buf);
        if (err !== null) return [s, err];
        s.zip = v;
    }
    return [s, null];
}

export const setSimInfo = (buf: _.Buffer, s: SimInfo): Error | null => {
    if (s === null || s === undefined) return new Error(`set SimInfo: value is null or undefined`);
    const startOffset = buf.write_offset;
    const bits = new Uint8Array(1);
    if (!_.eqU32(s.id, 0)) {
        _.SetBit(bits, 0, true);
    }
    if (!_.eqText(s.title, "")) {
        _.SetBit(bits, 1, true);
    }
    if (!_.eqText(s.content, "")) {
        _.SetBit(bits, 2, true);
    }
    _.SetBit(bits, 3, s.a as boolean);
    _.SetBit(bits, 4, s.b as boolean);
    _.SetBit(bits, 5, s.c as boolean);
    _.SetBit(bits, 6, s.d as boolean);
    if (!_.eqBin(s.zip, new Uint8Array(0))) {
        _.SetBit(bits, 7, true);
    }

    const errBits = buf.write(bits);
    if (errBits !== null) return errBits;
    if (!_.eqU32(s.id, 0)) {
        const err = _.setU32(buf, s.id);
        if (err !== null) {
            buf.rewindWrite(startOffset);
            return err;
        }
    }
    if (!_.eqText(s.title, "")) {
        const err = _.setText(buf, s.title);
        if (err !== null) {
            buf.rewindWrite(startOffset);
            return err;
        }
    }
    if (!_.eqText(s.content, "")) {
        const err = _.setText(buf, s.content);
        if (err !== null) {
            buf.rewindWrite(startOffset);
            return err;
        }
    }
    if (!_.eqBin(s.zip, new Uint8Array(0))) {
        const err = _.setBin(buf, s.zip);
        if (err !== null) {
            buf.rewindWrite(startOffset);
            return err;
        }
    }
    return null;
}

export const getSimInfoList = (buf: _.Buffer): [SimInfo[], Error | null] => {
    const [count, err] = _.getU16(buf);
    if (err !== null) return [[], err];
    const list: SimInfo[] = new Array(count);
    for (let i = 0; i < count; i++) {
        const [item, err2] = getSimInfo(buf);
        if (err2 !== null) return [[], err2];
        list[i] = item;
    }
    return [list, null];
}
export const setSimInfoList = (buf: _.Buffer, v: SimInfo[]): Error | null => {
    if (v.length > 65535) return new Error(`list length ${v.length} exceeds u16 max`);
    const err = _.setU16(buf, v.length);
    if (err !== null) return err;
    for (const item of v) {
        const err2 = setSimInfo(buf, item);
        if (err2 !== null) return err2;
    }
    return null;
}
export const eqSimInfoList = (a: SimInfo[], b: SimInfo[]): boolean => _.eqList(a, b, eqSimInfo);

File: go/sb/struct_sim_info.go
package sb

import (
	"bytes"
	"fmt"
	"slices"
)

type SimInfo struct {
	Id uint32 `bson:"id" json:"id"`
	Title string `bson:"title" json:"title"`
	Content string `bson:"content" json:"content"`
	A bool `bson:"a" json:"a"`
	B bool `bson:"b" json:"b"`
	C bool `bson:"c" json:"c"`
	D bool `bson:"d" json:"d"`
	Zip []byte `bson:"zip" json:"zip"`
}

func sizeSimInfoBody(s *SimInfo, bits []byte) (int, error) {
	if s == nil { return 0, fmt.Errorf("sizeSimInfo: nil value") }
	bodySize := 0
	if s.Id != 0 {
		bodySize += 4
		if bits != nil { SetBit(bits, uint8(0), true) }
	}
	if s.Title != "" {
		fieldSize, err := sizeText(s.Title)
		if err != nil { return 0, fmt.Errorf("sizeSimInfo Title: %w", err) }
		bodySize += fieldSize
		if bits != nil { SetBit(bits, uint8(1), true) }
	}
	if s.Content != "" {
		fieldSize, err := sizeText(s.Content)
		if err != nil { return 0, fmt.Errorf("sizeSimInfo Content: %w", err) }
		bodySize += fieldSize
		if bits != nil { SetBit(bits, uint8(2), true) }
	}
	if bits != nil { SetBit(bits, uint8(3), s.A) }
	if bits != nil { SetBit(bits, uint8(4), s.B) }
	if bits != nil { SetBit(bits, uint8(5), s.C) }
	if bits != nil { SetBit(bits, uint8(6), s.D) }
	if s.Zip != nil {
		fieldSize, err := sizeBin(s.Zip)
		if err != nil { return 0, fmt.Errorf("sizeSimInfo Zip: %w", err) }
		bodySize += fieldSize
		if bits != nil { SetBit(bits, uint8(7), true) }
	}
	return bodySize, nil
}

func sizeSimInfo(s *SimInfo) (int, error) {
	bodySize, err := sizeSimInfoBody(s, nil)
	if err != nil { return 0, err }
	return 1 + bodySize, nil
}

func sizeSimInfoList(v SimInfoList) (int, error) {
	if len(v) > 65535 { return 0, fmt.Errorf("list length exceeds uint16 max") }
	totalSize := 2
	for i, item := range v {
		itemSize, err := sizeSimInfo(item)
		if err != nil { return 0, fmt.Errorf("sizeSimInfoList[%d]: %w", i, err) }
		totalSize += itemSize
	}
	return totalSize, nil
}

func GetSimInfo(buf *bytes.Buffer, s *SimInfo) error {
	if s == nil { return nil }
	const bitSize = 1
	if buf.Len() < bitSize { return fmt.Errorf("GetSimInfo bitmask: %d - %d", buf.Len(), bitSize) }
	bits := buf.Next(bitSize)
	if GetBit(bits, uint8(0)) {
		val, err := GetU32(buf)
		if err != nil { return fmt.Errorf("GetSimInfo Id: %w", err) }
		s.Id = val
	}
	if GetBit(bits, uint8(1)) {
		val, err := GetText(buf)
		if err != nil { return fmt.Errorf("GetSimInfo Title: %w", err) }
		s.Title = val
	}
	if GetBit(bits, uint8(2)) {
		val, err := GetText(buf)
		if err != nil { return fmt.Errorf("GetSimInfo Content: %w", err) }
		s.Content = val
	}
	s.A = GetBit(bits, uint8(3))
	s.B = GetBit(bits, uint8(4))
	s.C = GetBit(bits, uint8(5))
	s.D = GetBit(bits, uint8(6))
	if GetBit(bits, uint8(7)) {
		val, err := GetBin(buf)
		if err != nil { return fmt.Errorf("GetSimInfo Zip: %w", err) }
		s.Zip = val
	}
	if err := ValidateSimInfo(s); err != nil { return fmt.Errorf("ValidateSimInfo: %w", err) }
	return nil
}

func SetSimInfo(buf *bytes.Buffer, s *SimInfo) error {
	if s == nil { return fmt.Errorf("SetSimInfo: nil value") }
	if err := ValidateSimInfo(s); err != nil { return fmt.Errorf("ValidateSimInfo: %w", err) }
	const bitSize = 1
	var bits [1]byte
	bodySize, err := sizeSimInfoBody(s, bits[:])
	if err != nil { return err }
	buf.Grow(bitSize + bodySize)
	if _, err := buf.Write(bits[:]); err != nil { return fmt.Errorf("SetSimInfo write bitmask: %w", err) }
	if s.Id != 0 {
		if err := SetU32(buf, s.Id); err != nil { return fmt.Errorf("SetSimInfo Id: %w", err) }
	}
	if s.Title != "" {
		if err := SetText(buf, s.Title); err != nil { return fmt.Errorf("SetSimInfo Title: %w", err) }
	}
	if s.Content != "" {
		if err := SetText(buf, s.Content); err != nil { return fmt.Errorf("SetSimInfo Content: %w", err) }
	}
	if s.Zip != nil {
		if err := SetBin(buf, s.Zip); err != nil { return fmt.Errorf("SetSimInfo Zip: %w", err) }
	}
	return nil
}

func ValidateSimInfo(s *SimInfo) error {
	if s == nil { return nil }
	return nil
}

func EqSimInfo(a, b *SimInfo) bool {
	if a == b { return true }
	if a == nil || b == nil { return false }
	if !EqU32(a.Id, b.Id) { return false }
	if !EqText(a.Title, b.Title) { return false }
	if !EqText(a.Content, b.Content) { return false }
	if !EqBool(a.A, b.A) { return false }
	if !EqBool(a.B, b.B) { return false }
	if !EqBool(a.C, b.C) { return false }
	if !EqBool(a.D, b.D) { return false }
	if !EqBin(a.Zip, b.Zip) { return false }
	return true
}

// Standalone functions
func ReadSimInfo(buf *bytes.Buffer) (*SimInfo, error) {
	s := new(SimInfo)
	return s, GetSimInfo(buf, s)
}

type SimInfoList []*SimInfo
func GetSimInfoList(buf *bytes.Buffer) (SimInfoList, error) {
	count, err := GetU16(buf)
	if err != nil { return nil, err }
	list := make([]*SimInfo, count)
	for i := range list {
		item, err := ReadSimInfo(buf)
		if err != nil { return nil, err }
		list[i] = item
	}
	return SimInfoList(list), nil
}
func SetSimInfoList(buf *bytes.Buffer, v SimInfoList) error {
	totalSize, err := sizeSimInfoList(v)
	if err != nil { return err }
	buf.Grow(totalSize)
	if err := SetU16(buf, uint16(len(v))); err != nil { return err }
	for _, item := range v {
		if err := SetSimInfo(buf, item); err != nil { return err }
	}
	return nil
}
func ValidateSimInfoList(v SimInfoList) error {
	for i, item := range v {
		if item == nil { return fmt.Errorf("SimInfoList[%d]: nil item", i) }
		if err := ValidateSimInfo(item); err != nil { return fmt.Errorf("SimInfoList[%d]: %w", i, err) }
	}
	return nil
}
func EqSimInfoList(a, b SimInfoList) bool { return slices.EqualFunc(a, b, EqSimInfo) }
